source/Calamari.Aws/Deployment/Conventions/CreateCloudFormationChangeSetConvention.cs

```diff
@@ -0,0 +1,137 @@
+            Guard.NotNullOrWhiteSpace(name, "The changeset name must be provided.");
+
+            var template = templateFactory();
+            Guard.NotNull(template, "CloudFormation template should not be null.");
+
+            try
+            {
+                var status = clientFactory.StackExists(stack, StackStatus.DoesNotExist);
+
+                CreateChangesetRequest(
+                    status,
+                    name,
+                    stack,
+                    roleArnProvider?.Invoke(deployment),
+                    template,
+                    Capabilities
+                )
+                .Map(CreateChangeSet)
+                .Tee(WaitForChangesetCompletion)
+                .Tee(ApplyVariables(deployment.Variables));
+            }
+            catch (AmazonServiceException exception)
+            {
+                HandleAmazonServiceException(exception);
+                throw;
+            }
+        }
+
+        private void WaitForChangesetCompletion(RunningChangeSet result)
+        {
+            clientFactory.WaitForChangeSetCompletion(CloudFormationDefaults.StatusWaitPeriod, result);
+        }
+
+        private Action<RunningChangeSet> ApplyVariables(CalamariVariableDictionary variables)
+        {
+            return result =>
+            {
+                SetOutputVariable(variables, "ChangesetId", result.ChangeSet.Value);
+                SetOutputVariable(variables, "StackId", result.Stack.Value);
+                variables.Set(AwsSpecialVariables.CloudFormation.Changesets.Arn, result.ChangeSet.Value);
+            };
+        }
+
+        private CreateChangeSetRequest CreateChangesetRequest(StackStatus status, string changesetName, StackArn stack, string roleArn, CloudFormationTemplate template, List<string> capabilities)
+        {
+            return new CreateChangeSetRequest
+            {
+                StackName = stack.Value,
+                TemplateBody = template.Content,
+                Parameters = template.Inputs.ToList(),
+                ChangeSetName = changesetName,
+                ChangeSetType = status == StackStatus.DoesNotExist ? ChangeSetType.CREATE : ChangeSetType.UPDATE,
+                Capabilities = capabilities,
+                RoleARN = roleArn 
+            };
+        }
+
+        private RunningChangeSet CreateChangeSet(CreateChangeSetRequest request)
+        {
+            try
+            {
+                return clientFactory.CreateChangeSet(request)
+                    .Map(x => new RunningChangeSet(new StackArn(x.StackId), new ChangeSetArn(x.Id)));
+            }
+            catch (AmazonCloudFormationException ex)
+            {
+                if (ex.ErrorCode == "AccessDenied")
+                {
+                    throw new PermissionException(
+                        @"AWS-CLOUDFORMATION-ERROR-0017: The AWS account used to perform the operation does not have " +
```

Comment:
Super minor, but it feels like these could be abstracted into some sort of template function that accepts: `errorNumber, errorSummary, exceptionMessage, includeDocTextAndLink`. Some of this just feels very duplicated, even though I know they're all slightly different based on error number

Reply:
Yeah, definitely agree there. This is partly what I was referring to reviewing error codes in https://github.com/OctopusDeploy/Issues/issues/4778. These error codes in general need to be revisited and possibly changed so that we aren't being so repetitious. There are a number of locations where the error codes probably should be the same as well such as the Permission Exception above. The error message should be enough to provide some context, right? :)